core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+                i == 0
+            })
+    }
+
+    // Puts by primary key first, then retrieves the existing value for each index and appends the
+    // newly created primary key value.
+    // TODO(ekmartin): This will put exactly the values that are given, and can only be retrieved
+    // with exactly those values. I think the regular state implementation supports inserting
+    // something like an Int and retrieving with a BigInt.
+    fn insert(&mut self, batch: &mut WriteBatch, r: Vec<DataType>) {
+        for index in self.indices.iter_mut() {
+            index.seq += 1;
+        }
+
+        let (secondary_indices, serialized_pk) = if let Some(ref pk_cols) = self.primary_key {
+            let pk = pk_cols.iter().map(|i| &r[*i]).collect::<Vec<_>>();
+            (&self.indices[..], Self::serialize_prefix(0, &pk))
+        } else {
+            // For bases without primary keys we store the actual row values keyed by the index
+            // that was added first. This means that we can't consider the keys unique though, so
+            // we'll append a sequence number.
+            let pk_index = &self.indices[0];
+            let pk = pk_index.columns.iter().map(|i| &r[*i]).collect::<Vec<_>>();
+            (&self.indices[1..], self.serialize_key(0, &pk, pk_index.seq))
+        };
+
+        // First insert the actual value for our primary index:
+        let pk_value = bincode::serialize(&r).unwrap();
+        batch.put(&serialized_pk, &pk_value).unwrap();
+
+        // Then insert primary key pointers for all the secondary indices:
+        for (i, index) in secondary_indices.iter().enumerate() {
+            // Construct a key with the index values, and serialize it with bincode:
+            let key = index.columns.iter().map(|i| &r[*i]).collect::<Vec<_>>();
+            // Add 1 to i since we're slicing self.indices by 1..:
+            let serialized_key = self.serialize_key((i + 1) as u32, &key, index.seq);
+            batch.put(&serialized_key, &serialized_pk).unwrap();
+        }
+    }
+
+    fn remove(&self, batch: &mut WriteBatch, r: &[DataType]) {
+        let db = self.db.as_ref().unwrap();
+        let mut delete_cols: Vec<_> = self.indices
+            .iter()
+            .enumerate()
+            .map(|(i, index)| {
+                if self.primary_key.is_some() {
+                    (i + 1, &index.columns)
+                } else {
+                    (i, &index.columns)
+                }
+            })
+            .collect();
+
+        if let Some(ref pk_cols) = self.primary_key {
+            delete_cols.push((0, pk_cols));
+        }
+
+        for (i, columns) in delete_cols.into_iter() {
+            let index_row = columns.iter().map(|i| &r[*i]).collect::<Vec<_>>();
+            let serialized_key = Self::serialize_prefix(i as u32, &index_row);
+            for (key, _value) in db.prefix_iterator(&serialized_key) {
+                batch.delete(&key).unwrap();
```

Comment:
This doesn't seem right. `remove` should remove *one*, not *all* rows. And specifically should only remove one row that *exactly matches* `r`, not just where the key is the same.. And I suspect that means you have to remove one row from the first index first, and then search the secondary indices for the entry that references the removed row from the first index.

Reply:
Nice catch! Fixed in a3e9e1fbccd2bd4f5cacb5aa45e494d41d63cb78.